synapse/rest/admin/rooms.py

```diff
@@ -338,3 +341,97 @@ async def on_POST(self, request, room_identifier):
         )
 
         return 200, {"room_id": room_id}
+
+
+class MakeRoomAdminRoomServlet(RestServlet):
+    PATTERNS = admin_patterns("/make_room_admin/(?P<room_identifier>[^/]*)")
+
+    def __init__(self, hs: "HomeServer"):
+        self.hs = hs
+        self.auth = hs.get_auth()
+        self.room_member_handler = hs.get_room_member_handler()
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.state_handler = hs.get_state_handler()
+        self.is_mine_id = hs.is_mine_id
+
+    async def on_POST(self, request, room_identifier):
+        requester = await self.auth.get_user_by_req(request)
+        await assert_user_is_admin(self.auth, requester.user)
+        content = parse_json_object_from_request(request, allow_empty_body=True)
+
+        if RoomID.is_valid(room_identifier):
+            room_id = room_identifier
+        elif RoomAlias.is_valid(room_identifier):
+            room_alias = RoomAlias.from_string(room_identifier)
+            room_id, _ = await self.room_member_handler.lookup_room_alias(room_alias)
+            room_id = room_id.to_string()
+        else:
+            raise SynapseError(
+                400, "%s was not legal room ID or room alias" % (room_identifier,)
+            )
+
+        user_to_add = content.get("user_id", requester.user.to_string())
+
+        room_state = await self.state_handler.get_current_state(room_id)
+
+        if not room_state:
+            raise SynapseError(400, "Server not in room")
+
+        create_event = room_state[(EventTypes.Create, "")]
+        power_levels = room_state.get((EventTypes.PowerLevels, ""))
+
+        if power_levels is not None:
+            admin_users = [
+                user_id
+                for user_id, power in power_levels.content.get("users")
+                if self.is_mine_id(user_id)
+            ]
+            admin_users.sort(key=lambda user: power_levels.content.get("users")[user])
+
+            admin_user_id = admin_users[-1]
```

Comment:
what if there are no local users with a high enough PL to do this operation?

Reply:
@erikjohnston I think this is an outstanding comment?

Reply:
D'oh, forgot to respond.

This has also changed. We try and change the power levels with the local user with the maximal power level, and if that fails with an `AuthError` we convert it to a 400 (and there is even a test for this)